src/components/TagContainer.ts

```diff
@@ -0,0 +1,277 @@
+            return readOnly;
+        }
+
+        return true;
+    }
+
+    private resetSubscriptions(mxObject?: mendix.lib.MxObject) {
+        this.subscriptionHandles.forEach(window.mx.data.unsubscribe);
+
+        if (mxObject) {
+            this.subscriptionHandles.push(mx.data.subscribe({
+                callback: guid => {
+                    mx.data.get({
+                        guid: guid.toString(),
+                        callback: object => this.fetchTags(object)
+                    });
+                },
+                guid: mxObject.getGuid()
+            }));
+
+            this.subscriptionHandles.push(mx.data.subscribe({
+                attr: this.referenceAttribute,
+                callback: guid => {
+                    mx.data.get({
+                        guid: guid.toString(),
+                        callback: object => this.fetchTags(object)
+                    });
+                },
+                guid: mxObject.getGuid()
+            }));
+
+            this.subscriptionHandles.push(mx.data.subscribe({
+                guid: mxObject.getGuid(),
+                val: true,
+                callback: validations => {
+                    window.mx.ui.error(validations[0].getErrorReason(this.referenceAttribute));
+                }
+            }));
+        }
+    }
+
+    private lazyLoadSuggestions() {
+        this.setState({ lazyLoaded: true });
+        this.fetchTags(this.props.mxObject);
+    }
+
+    private fetchTags(mxObject: mendix.lib.MxObject) {
+        if (!this.state.fetchTags) {
+            const { tagConstraint } = this.props;
+            const constraint = tagConstraint
+                ? tagConstraint.replace(/\[\%CurrentObject\%\]/gi, mxObject.getGuid())
+                : "";
+            const XPath = "//" + this.tagEntity + constraint;
+
+            mx.data.get({
+                callback: object => this.processTags(object),
+                error: error =>
+                    window.mx.ui.error(`An error occurred while retrieving tags via XPath (${this.tagEntity}):
+                ${error.message}`),
+                xpath: XPath
+            });
+        }
+    }
+
+    private processTags(tagObjects: mendix.lib.MxObject[]) {
+        const currentTags: mendix.lib.MxObject[] = [];
+        const referenceTags = this.props.mxObject.getReferences(this.referenceAttribute) as string[];
+        const getSuggestions = tagObjects.map(object => ({ value: object.get(this.props.tagAttribute) as string }));
+
+        tagObjects.map(object => {
+            if (referenceTags.toString() !== "") {
+                referenceTags.map(reference => {
+                    if (reference === object.getGuid()) {
+                        currentTags.push(object);
+                    }
+                });
+            }
+        });
+        const getTags = currentTags !== []
+        ? currentTags.map(object => ({ value: object.get(this.props.tagAttribute) as string }))
+        : [];
```

Comment:
wrong indenting
``` ts
const variable = expression ?
   ? trueValue
   : valseValue;
```
Though there is no need to do it al all. `.map` will not fail on an empty array but will just return and empty array
 ``` ts
const getTags = currentTags.map(object => ({ value: object.get(this.props.tagAttribute) as string }))
```
name `getTags` does not feel like a variable name, more like a function name..

though this code does not make sens too as ou extract the tag again on 187. you can do at once
``` ts
const tagList: string[] = currentTags.map(object => object.get(this.props.tagAttribute) as string);
```
Same count for `suggestions`